include/bout/indexoffset.hxx

```diff
@@ -0,0 +1,50 @@
+#ifndef __INDEXOFFSET_H__
+#define __INDEXOFFSET_H__
+
+#include "bout/mesh.hxx"
+#include "bout/region.hxx"
+
+template <typename T = Ind3D>
+struct IndexOffset {
+
+  static_assert(std::is_base_of<Ind2D, T>::value || std::is_base_of<Ind3D, T>::value,
+                "IndexOffset must be templated with either Ind2D or Ind3D");
+
+  Mesh &mesh;
+  const int nx, ny, nz;
+
+  IndexOffset(Mesh &mesh)
+      : mesh(mesh), nx(mesh.LocalNx), ny(mesh.LocalNy), nz(mesh.LocalNz) {}
+
+  /// Convenience functions for converting to (x, y, z)
+  int x(T index) const { return (index.ind / nz) / ny; }
+  int y(T index) const { return (index.ind / nz) % ny; }
+  int z(T index) const { return (index.ind % nz); }
+
+  const inline T xp(T index, int i = 1) const { return index + (i * ny * nz); }
+  /// The index one point -1 in x
+  const inline T xm(T index, int i = 1) const { return index - (i * ny * nz); }
+  /// The index one point +1 in y
+  const inline T yp(T index, int i = 1) const { return index + (i * nz); }
+  /// The index one point -1 in y
+  const inline T ym(T index, int i = 1) const { return index - (i * nz); }
+  /// The index one point +1 in z. Wraps around zend to zstart
+  const inline T zp(T index, int i = 1) const {
+    ASSERT3(i > 0);
+    return (index + i) % nz == 0 ? index - nz + i : index + i;
```

Comment:
We're paying the cost of a mod for each call and then doing another calculation to find the result. I think we can probably just use a calculation involving mod to find the answer without a further calculation? The ternary style is handy when we know i can only be 1 so that we can avoid the mod in the logic part.

Reply:
We discussed this earlier, but for future reference: because we only have the single index, we have to pay the cost of the modulus to check if we've reached either end of z, but the modulus is not sufficient for the actual periodicity -- we need to shift by nz.

Hopefully, the z-offsetting is not performance critical. If it is, we could switch to using precalculated offset indices. This interface should still work in that case.